primitives_derive/src/migrate.rs

```diff
@@ -126,31 +132,41 @@ enum MigrateRefs {
     Any,
     /// Derived from `#[migrate(ident, ident, ...)]`.
     /// Only those identifiers in the list and which match in the type of the field should be migrated.
-    Listed(Vec<syn::Ident>),
+    Listed(Vec<Ident>),
+    /// Do an exact replacement of the field type with the one given in `#[migrate_from(Type)]`.
+    Exact(Type),
 }
 
-/// Returns information about any `#[migrate]` attributes while also stripping them.
+/// Returns information about any `#[migrate]` or `#[migrate_from]` attributes.
+/// We also strip those attributes while at it.
 ///
 /// The form `#[migrate = ".."]` does qualify.
 fn extract_migrate_refs(attrs: &mut Vec<syn::Attribute>) -> Option<MigrateRefs> {
     let mut mig_ref = None;
     attrs.retain(|attr| {
-        // Only care about `migrate], and remove all of those, irrespective of form.
+        // Only care about `migrate`, and remove all of those, irrespective of form.
         if attr.path.is_ident("migrate") {
-            if attr.tokens.is_empty() {
+            mig_ref = Some(if attr.tokens.is_empty() {
                 // Got exactly `#[migrate]`.
                 // User doesn't wish to specify which types to migrate, so assume all.
-                mig_ref = Some(MigrateRefs::Any);
-            } else if let Ok(refs) = attr.parse_args_with(|ps: ParseStream| {
-                // Got `migrate(ident, ident, ...)`.
-                // User only wants to oldify the given identifiers.
-                // Applies in e.g., `field: Vec<Foo>` where `Foo` is being migrated
-                // but `Vec` shouldn't be renamed as it is a container of `Foo`s.
-                ps.parse_terminated::<_, syn::Token![,]>(Ident::parse)
-                    .map(|iter| iter.into_iter().collect())
-            }) {
-                mig_ref = Some(MigrateRefs::Listed(refs));
-            }
+                MigrateRefs::Any
+            } else {
+                MigrateRefs::Listed(
+                    attr.parse_args_with(|ps: ParseStream| {
+                        // Got `migrate(ident, ident, ...)`.
+                        // User only wants to oldify the given identifiers.
+                        // Applies in e.g., `field: Vec<Foo>` where `Foo` is being migrated
+                        // but `Vec` shouldn't be renamed as it is a container of `Foo`s.
+                        ps.parse_terminated::<_, syn::Token![,]>(Ident::parse)
+                            .map(|iter| iter.into_iter().collect())
+                    })
+                    .unwrap(),
+                )
+            });
+            false
+        } else if attr.path.is_ident("migrate_from") {
```

Comment:
Instead of repeated string equality checks, can you `get_ident` and then do a `match` on the result?

Reply:
Sure; Adjusted this in https://github.com/PolymathNetwork/Polymesh/pull/592/commits/3b9e89f262db646cd66126fb58a8050420533728.

Unfortunately this does still require some allocation due to the way the API works, but it might have already allocated under the hood anyways. It's always difficult to tell with proc macros whether a certain change will improve compile times or not, since making the proc macro itself more complicated may dominate the compile times as compared to the macro invocations. However, I think your suggestion improves readability, so I've implemented it. :)

Reply:
Thanks! It's definitely more readable now. I particularly like the `Some("migrate") if attr.tokens.is_empty()` clause.